### ISSUE TITLE
[MNG-4761] Plugin-level dependency scope causes some plugin classpaths to be incorrect

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6337]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** opened **[MNG-4761](https://issues.apache.org/jira/browse/MNG-4761?redirect=false)** and commented

Plugin-level dependencies should use RUNTIME scope at all times. Using any other scope may alter the weighting given to the subgraph-choice algorithm used in transitive dependency resolution.

Plugin-level dependencies use compile scope by default. When transitive resolution takes place, compile scope takes precedence over runtime scope, causing the transitive dependency sub-graph of the plugin-level dependency to be activated over those of the plugin itself.

This happens even when the plugin's transitive dep is NEARER to the top level than the one brought in by that plugin-level dependency itself.

The result is that when a dep that's farther away is chosen over a nearer one, it can then be disabled by Maven choosing to disable its parent dep (the one that brought it in) in another part of the transitive resolution process.

This is a very subtle case where Maven is doing the wrong thing. The attached test case should make it clearer.

---

**Affects:** 2.2.1, 3.0-beta-1, 3.0-beta-2

**Attachments:**
- [MNG-4761-mvn3b2.patch](https://issues.apache.org/jira/secure/attachment/12712396/MNG-4761-mvn3b2.patch) (_6.92 kB_)
- [MNG-4761-mvn3b2.reformatted.patch](https://issues.apache.org/jira/secure/attachment/12713241/MNG-4761-mvn3b2.reformatted.patch) (_829 bytes_)
- [obscured-nearer-dep.zip](https://issues.apache.org/jira/secure/attachment/12713072/obscured-nearer-dep.zip) (_38.79 kB_)

**Issue Links:**
- [MNG-4790](https://issues.apache.org/jira/browse/MNG-4790) Dependency resolution issue with provided scope

- [MNG-4800](https://issues.apache.org/jira/browse/MNG-4800) Conflict resolution does not pick sub tree of nearest dependency if farther conflicting dependency has wider scope

- [MNG-4134](https://issues.apache.org/jira/browse/MNG-4134) Conflict resolution with scope widening can cause transitive dependencies to be omitted

**Backported to:** [2.2.2](https://github.com/apache/maven/milestone/56?closed=1)

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

Patch to force all plugin-level dependencies to use scope: runtime. We could probably use a debug statement showing this scope being managed to runtime, but the patch works.

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

Fixed to remove unrelated reformats.

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

all scopes EXCEPT system will be forced to runtime for plugin-level dependencies.

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

For future reference, the failing test scenario centeres around the following dirty tree for the plugin depencencies:

```
plugin
+- override:0.1:compile
|  \- middle:0.1:compile        <-----------\
|     \- required:0.1:compile    <---\      |
+- direct:0.1:runtime                (a)    |
|  +- required:0.1:runtime       <---/      |
|  \- middle:0.1:runtime        <--------- (b)
|     \- required:0.1:runtime               |
\- middle:0.2:runtime           <-----------/
```

During resolution of conflict (a), the node `required:0.1:runtime` will be disabled in favor of `required:0.1:compile`, although the runtime-scope dependency is nearer, i.e. Maven's usual conflict resolution strategy of nearest-wins is violated here. When conflict (b) is resolved, all `middle:0.1:*` nodes including their children will be disabled, thereby completely losing `required:0.1` on the plugin class path.

So while using runtime scope for the project-level plugin dependencies workarounds the problem, the issue remains for ordinary project dependency resolution. I wonder if it got filled already.
